src/main/kotlin/com/github/avrokotlin/avro4k/Avro.kt

```diff
@@ -15,266 +15,322 @@ import kotlinx.serialization.BinaryFormat
+     */
+    @Deprecated("Has been replaced by #encode that is able to not-only encode records", replaceWith = ReplaceWith("this.encode(serializer, obj)"))
+    fun <T> toRecord(
+        serializer: SerializationStrategy<T>,
+        obj: T,
+    ): GenericRecord {
+        return toRecord(serializer, schema(serializer.descriptor), obj)
+    }
+
+    /**
+     * Converts an instance of <T> to an Avro [Record] using the given [Schema].
+     */
+    @Deprecated("Has been replaced by #encode that is able to encode everything", replaceWith = ReplaceWith("this.encode(serializer, schema, obj)"))
+    fun <T> toRecord(serializer: SerializationStrategy<T>,
+                     schema: Schema,
+                     obj: T): GenericRecord {
+        val encodedValue = encodeToGenericData(serializer, schema, obj)
+        if (encodedValue !is GenericRecord)
+            throw AvroRuntimeException("Expected a GenericRecord, found $encodedValue")
+        return encodedValue
+    }
+
+    fun <T> encodeToGenericData(serializer: SerializationStrategy<T>, schema: Schema, value: T): Any? =
+        GenericAvroEncoder(schema, serializersModule, configuration).apply {
+            resolveSchema(serializer.descriptor, value == null)
+            encodeSerializableValue(serializer, value)
+        }.encodedValue
+
+    /**
+     * Converts an Avro [GenericRecord] to an instance of <T> using the schema
+     * present in the record.
+     */
+    @Deprecated("Has been replaced by #decodeFromGenericData that is able to decode everything", replaceWith = ReplaceWith("this.decodeFromGenericData(deserializer, record)"))
+    fun <T> fromRecord(
+        deserializer: DeserializationStrategy<T>,
+        record: GenericRecord,
+    ): T {
+        return GenericAvroDecoder(record, record.schema, serializersModule, configuration)
+            .decodeSerializableValue(deserializer)
+    }
+
+    fun <T> decodeFromGenericData(
+        deserializer: DeserializationStrategy<T>,
+        schema: Schema,
+        value: Any?,
+    ): T? {
+        return GenericAvroDecoder(value, schema, serializersModule, configuration)
+            .decodeNullableSerializableValue(deserializer)
+    }
+
+    fun schema(descriptor: SerialDescriptor): Schema = schemaFor(
+        serializersModule,
+        descriptor,
+        descriptor.annotations,
+        configuration,
+        mutableMapOf()
+    ).schema()
+
+    fun <T> schema(serializer: SerializationStrategy<T>): Schema =
+        schema(serializer.descriptor)
+}
+
+inline fun <reified T> Avro.encodeToGenericData(value: T): Any? {
+    val serializer = serializersModule.serializer<T>()
+    return encodeToGenericData(serializer, schema(serializer.descriptor), value)
+}
+
+fun <T> Avro.encodeToGenericData(serializer: SerializationStrategy<T>, value: T): Any? =
+    encodeToGenericData(serializer, schema(serializer.descriptor), value)
+
+inline fun <reified T> Avro.encodeToGenericData(schema: Schema, value: T) =
```

Comment:
how about merging it with the other reified encodeToGenericData method?
```kotlin
inline fun <reified T> Avro.encodeToGenericData(value: T, serializer: SerializationStrategy<T> = serializersModule.serializer<T>(), schema: Schema = schema(serializer.descriptor)): Any? {
    return encodeToGenericData(value, serializer, schema)
}
```
I would change the signature of the encodeToGenericData method in the same way that the parameters have the following order: value, serializer, schema.

That way users can use the API in the following way:
```kotlin
Avro.default.encodeToGenericData("blu", serializer, schema)
Avro.default.encodeToGenericData("blu")
```

Reply:
Good point, I just reused the same as before if I remember

Reply:
I'll put the serialization strategy at the end as it is not required in most cases, so in this order: value, schema, serializer

Reply:
I can't, so let's go in your order

Reply:
The issue is that we have this error now: Extension is shadowed by a member. I don't have easy solution.